File: ui/src/checkers/helm_config/webapp_backend_urls.ts
/**
 * Checks whether webapp backend URLs have been changed from placeholder values.
 *
 * Consumes the config/webapp_backend_urls target (boolean or string).
 * Placeholder URLs like «api.example.com» are the #1 cause of
 * «ERROR 6» after a fresh Wire install.
 */

// Ours
import { BaseChecker, type CheckResult } from '../base_checker'
import type { DataLookup } from '../data_lookup'

export class WebappBackendUrlsChecker extends BaseChecker {
    readonly path: string = 'helm_config/webapp_backend_urls'
    readonly name: string = 'Webapp backend URLs not placeholder values'
    readonly category: string = 'Helm / Config Validation'
    readonly interest = 'Setup' as const
    readonly explanation: string = 'Detects **placeholder** backend URLs (like `api.example.com`) in the webapp configuration. Unreplaced placeholders are the most common cause of **"ERROR 6"** after a fresh Wire install.'

    check(data: DataLookup): CheckResult {
        const point = data.get('config/webapp_backend_urls')

        // Target data was not collected
        if (!point) {
            return {
                status: 'gather_failure',
                status_reason: 'Webapp backend URL data was not collected.',
                fix_hint: '1. Verify the webapp ConfigMap exists:\n   ```\n   kubectl get configmap webapp -n wire -o yaml\n   ```\n2. Re-run the gatherer ensuring the `config/webapp_backend_urls` target succeeds.',
                recommendation: 'Couldn\'t collect webapp backend URL data.',
            }
        }

        const val: string | boolean = point.value as string | boolean

        // New structured return values from target: "ok", "error", "warning"
        if (val === 'ok') {
            return {
                status: 'healthy',
                status_reason: 'Webapp backend URLs are properly configured (no placeholder patterns detected).',
                display_value: 'configured',
                raw_output: point.raw_output,
            }
        }

        if (val === 'error') {
            return {
                status: 'unhealthy',
                status_reason: 'Webapp backend URLs contain placeholder values.',
                recommendation: 'Webapp config still points to placeholder URLs (e.g., api.example.com, CHANGE_ME). This is the most common cause of «ERROR 6» after a fresh install. Replace them with the actual backend service URLs.',
                display_value: 'placeholder URLs',
                raw_output: point.raw_output,
            }
        }

        if (val === 'warning') {
            return {
                status: 'warning',
                status_reason: 'Webapp backend URLs reference localhost/127.0.0.1.',
                recommendation: 'Localhost URLs may be valid for single-node or co-located deployments (e.g., local Redis sidecar). Verify this is intentional — if this is a multi-node deployment, these should be replaced with real service addresses.',
                display_value: 'localhost URLs',
                raw_output: point.raw_output,
            }
        }

        // Legacy boolean support (old gatherer data)
        if (val === true) {
            return {
                status: 'healthy',
                status_reason: 'Webapp backend URLs are properly configured (not placeholders).',
                display_value: val,
                raw_output: point.raw_output,
            }
        }

        // Boolean false — old gatherer flagged placeholder URLs
        return {
            status: 'unhealthy',
            status_reason: 'Webapp backend URLs are still set to placeholder values.',
            recommendation: 'Webapp config still points to placeholder URLs (e.g., api.example.com). This is the most common cause of «ERROR 6» after a fresh install.',
            display_value: val,
            raw_output: point.raw_output,
        }
    }
}

export default WebappBackendUrlsChecker

File: src/targets/config/webapp_backend_urls.py
"""Checks that webapp config doesn't still point to placeholder URLs.

This is one of the most common reasons for « ERROR 6 » after a fresh
install. Verifies the webapp ConfigMap doesn't contain example.com or
other placeholder values.
"""

from __future__ import annotations

# External
from typing import Any

# Ours
from src.lib.base_target import BaseTarget


# Patterns that definitely indicate unconfigured placeholder values
_ERROR_PATTERNS: list[str] = [
    "example.com",
    "placeholder",
    "CHANGE_ME",
    "changeme",
]

# Patterns that might be legitimate in single-node or co-located deployments
# (e.g., local Redis sidecar, co-located Elasticsearch), but could also
# indicate an incomplete configuration
_WARNING_PATTERNS: list[str] = [
    "localhost",
    "127.0.0.1",
]


class WebappBackendUrls(BaseTarget):
    """Checks webapp config for placeholder backend URLs.

    Fetches the webapp deployment or ConfigMap and makes sure backend
    URLs aren't still set to placeholder values.
    """

    @property
    def description(self) -> str:
        """What we're checking."""
        return "Webapp backend URLs not placeholder values"

    @property
    def explanation(self) -> str:
        """Why we're checking and what's healthy vs unhealthy."""
        return (
            "Placeholder URLs (example.com, CHANGE_ME) left in webapp config cause "
            "« ERROR 6 » on login. Localhost/127.0.0.1 URLs are flagged as warnings "
            "since they may be valid in single-node or co-located deployments."
        )

    @property
    def unit(self) -> str:
        """Unit of measurement (empty, result is boolean)."""
        return ""

    def collect(self) -> str | None:
        """Check webapp config for placeholder URLs.

        Returns:
            "ok" if no placeholder patterns found,
            "error" if definite placeholder patterns found (example.com, CHANGE_ME),
            "warning" if only localhost/127.0.0.1 found (may be legitimate),
            None if the webapp configuration could not be retrieved.
        """
        self.terminal.step("Checking webapp backend URLs...")

        # Try to get the webapp ConfigMap or deployment env vars
        _result, cm_data = self.run_kubectl("configmap/webapp")

        config_text: str = ""

        if isinstance(cm_data, dict):
            # Join all data keys together so we can search them
            data_section: dict[str, str] = cm_data.get("data", {})
            config_text = "\n".join(data_section.values())

        # Also check the deployment environment variables
        if not config_text:
            _result2, deploy_data = self.run_kubectl("deployment/webapp")
            if isinstance(deploy_data, dict):
                containers: list[dict[str, Any]] = (
                    deploy_data.get("spec", {})
                    .get("template", {})
                    .get("spec", {})
                    .get("containers", [])
                )
                for container in containers:
                    for env in container.get("env", []):
                        env_val: str = str(env.get("value", ""))
                        config_text += f"\n{env.get('name', '')}={env_val}"

        if not config_text:
            self._health_info = "Could not retrieve webapp configuration"
            return None  # Couldn't run the check

        config_lower: str = config_text.lower()

        # Check for definite placeholder patterns first
        found_errors: list[str] = []
        for pattern in _ERROR_PATTERNS:
            if pattern.lower() in config_lower:
                for line in config_text.split("\n"):
                    if pattern.lower() in line.lower():
                        found_errors.append(line.strip()[:80])
                        break

        if found_errors:
            self._health_info = (
                f"Placeholder URLs found: {'; '.join(found_errors[:3])}"
            )
            return "error"

        # Check for localhost/loopback patterns (may be legitimate)
        found_warnings: list[str] = []
        for pattern in _WARNING_PATTERNS:
            if pattern.lower() in config_lower:
                for line in config_text.split("\n"):
                    if pattern.lower() in line.lower():
                        found_warnings.append(line.strip()[:80])
                        break

        if found_warnings:
            self._health_info = (
                f"Localhost URLs found (may be valid): "
                f"{'; '.join(found_warnings[:3])}"
            )
            return "warning"

        self._health_info = "No placeholder URLs detected"
        return "ok"
